trieste/acquisition/function.py

```diff
@@ -528,15 +532,187 @@ def prepare_acquisition_function(
+        return mean[..., None, :] + tf.sqrt(cov[..., None, :]) * tf.cast(self._eps, cov.dtype)  # [..., S, L]
+
+
+class BatchReparametrizationSampler:
+    r"""
+    This sampler employs the *reparameterization trick* to approximate batches of samples from a
+    :class:`ProbabilisticModel`\ 's predictive joint distribution as
+
+    .. math:: x \mapsto \mu(x) + \epsilon L(x)
+
+    where :math:`L` is the Cholesky factor s.t. :math:`LL^T = \sigma`, and
+    :math:`\epsilon \sim \mathcal N (0, 1)` is constant for a given sampler, thus ensuring samples
+    form a continuous curve.
+    """
+    def __init__(self, sample_size: int, model: ProbabilisticModel):
+        """
+        :param sample_size: The number of samples for each batch of points. Must be positive.
+        :param model: The model to sample from.
+        :raise ValueError (or InvalidArgumentError): If ``sample_size`` is not positive.
+        """
+        tf.debugging.assert_positive(sample_size)
+
+        self._sample_size = sample_size
+        self._eps = tf.Variable(
+            tf.reshape([], [0, 0, sample_size]), shape=[None, None, sample_size], dtype=tf.float64,
+        )  # [L, 0, S]
+        self._model = model
+
+    def sample(self, at: BatchQueryPoints, jitter: float = DEFAULTS.JITTER) -> tf.Tensor:
+        """
+        Return approximate samples from the `model` specified at :meth:`__init__`. Multiple calls to
+        :meth:`sample`, for any given :class:`BatchReparametrizationSampler` and ``at``, will
+        produce the exact same samples. Calls to :meth:`sample` on *different*
+        :class:`BatchReparametrizationSampler` instances will produce different samples.
+
+        :param at: Batches of query points at which to sample the predictive distribution.
+        :param jitter: The size of the jitter to use when stabilising the Cholesky decomposition of
+            the covariance matrix.
+        :return: The samples, of shape `[..., S, B, L]`, where `S` is the `sample_size`, `B` the
+            number of points per batch, and `L` the number of latent model dimensions.
+        :raise ValueError (or InvalidArgumentError): If ``at`` is a scalar, or the (inferred) batch
+            size is not positive.
+        """
+        tf.debugging.assert_rank_at_least(at, 2)
+        batch_size = at.shape[-2]
+        tf.debugging.assert_positive(batch_size, "Batch size must be positive")
+
+        mean, cov = self._model.predict_latent_full_cov(at)  # [..., B, L], [..., L, B, B]
+
+        if tf.size(self._eps) == 0:
+            self._eps.assign(
+                tf.random.normal(
+                    [mean.shape[-1], batch_size, self._sample_size], dtype=tf.float64
+                )  # [L, B, S]
+            )
+
+        identity = tf.eye(batch_size, dtype=cov.dtype)  # [B, B]
+        cov_cholesky = tf.linalg.cholesky(cov + jitter * identity)  # [..., L, B, B]
+
+        variance_contribution = cov_cholesky @ tf.cast(self._eps, cov.dtype)  # [..., L, B, S]
```

Comment:
gpflow uses something different - probably for a good reason? We need to get someone's eyes on this.

Reply:
what do you mean by "uses something different"?

Reply:
`        jittermat = (
            tf.eye(D, batch_shape=mean_shape[:-1], dtype=default_float()) * default_jitter()
        )  # [..., N, D, D]
        eps_shape = tf.concat([mean_shape, [S]], 0)
        eps = tf.random.normal(eps_shape, dtype=default_float())  # [..., N, D, S]
        chol = tf.linalg.cholesky(cov + jittermat)  # [..., N, D, D]
        samples = mean[..., None] + tf.linalg.matmul(chol, eps)  # [..., N, D, S]
        samples = leading_transpose(samples, [..., -1, -3, -2])  # [..., S, N, D]`

Reply:
can you be more specific? I've refactored it a little for clarity, replaced `tf.linalg.matmul` with `@`, dropped the `batch_shape` cos I think broadcasting does it for you. Was there something in particular?